data/nat/choose.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2018 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+Mostly based on Jeremy Avigad's choose file in lean 2
+-/
+
+
+import data.nat.basic
+
+open nat
+def choose : ℕ → ℕ → ℕ
+| _              0  := 1
+| 0        (succ k) := 0
+| (succ n) (succ k) := choose n k + choose n (succ k)
+
+@[simp]
+lemma choose_zero_right (n : ℕ) : choose n 0 = 1 := by cases n; refl
+
+@[simp]
+lemma choose_zero_succ (k : ℕ) : choose 0 (succ k) = 0 := rfl
+
+lemma choose_succ_succ (n k : ℕ) : choose (succ n) (succ k) = choose n k + choose n (succ k) := rfl
+
+lemma choose_eq_zero_of_lt {n : ℕ} : ∀ {k}, n < k → choose n k = 0 :=
+begin
```

Comment:
This can be written very nice using the equation compiler:
```lean
lemma choose_eq_zero_of_lt : ∀ {n k}, n < k → choose n k = 0
| 0     (k+1) hk := by simp
| (n+1) (k+1) hk := 
  have hnk : n < k, from lt_of_succ_lt_succ hk,
  have hnk1 : n < k + 1, from lt_of_succ_lt hk,
  by rw [choose_succ_succ, choose_eq_zero_of_lt hnk, choose_eq_zero_of_lt hnk1]
```
The equation compiler takes care of the `absurd` cases (`n < m` is `n + 1 ≤ m` which is inductively defined). My personal preference is to state proofs explicitly using `have ..` instead of adding the proof to to the `rw` list.